lib/package_provider/cached_repository.rb

```diff
@@ -0,0 +1,35 @@
+module PackageProvider
+  # Class representing cached repository to avoid multiple cloninng
+  class CachedRepository < PackageProvider::Repository
+    class CloneInProgress < StandardError
+    end
+
+    def repo_cache_dir(treeish, paths, use_submodules)
+      @sha256 ||= Digest::SHA256.new
+      h = { treeish: treeish, paths: paths, submodule: use_submodules }
+
+      digest = sha256.hexdigest h.to_json
+
+      File.join(PackageProvider.config.repository_cache_root, digest)
+    end
+
+    def repo_ready?(path)
```

Comment:
probably we will need locking file as "#{path}.clonne_lock" as well. Otherwise several threads/process could start to clone simultaneously.